### ISSUE TITLE
Multiple validations per epoch with PytorchLightning

### ISSUE THREAD

**Issue by LucaMarconato** [optuna/optuna#2932]
With large datasets it sometimes makes sense to check the validation set before the epoch ends, because the model could reach decent performance even with just 1-2 epochs, so we want to monitor it early. 

This is exactly my scenario. When performing more than one validation per epoch I get this warning message from optuna:

```
.../site-packages/optuna/trial/_trial.py:590: UserWarning: The reported value is ignored because this `step` 1 is already reported
```

## Expected behavior

Every value should be reported; in the case in which Pytorch Lightning is used (like in my case), a better indicator for the progress of the training is not `epoch` but `step` (1 step = 1 forward pass).

Even more importantly than just ignoring to report a value, it should be checked that this behavior is not interfering with the hyperparameters optimization logic.

## Environment
- Optuna version: 2.9.1
- Python version: 3.8.5
- OS: Linux-3.10.0-1127.18.2.el7.x86_64-x86_64-with-glibc2.10
- (Optional) Other libraries and their versions:
  - pytorch-lightning: 1.4.5 

## Error messages, stack traces, or logs
(reporting the one written above)
```
.../site-packages/optuna/trial/_trial.py:590: UserWarning: The reported value is ignored because this `step` 1 is already reported
```

## Steps to reproduce

Run optuna on a model where the validation phase is done multiple times per epoch. For instance train any model with PyTorch Lightning with the following trainer:

```
 trainer = pl.Trainer(
        gpus=1,
        max_epochs=1,
        callbacks=[
            PyTorchLightningPruningCallback(trial, monitor="val_accuracy"),  # set your metric here
        ],
        num_sanity_val_steps=0,
        val_check_interval=1,  # an epoch will be many steps, so here we are logging multiple times per epoch
    )
```

**Comment by nzw0301** [comment]
I'm afraid that I've replaced the label with `feature` since this is not a bug, I suppose.

### Workaround

Technically, it is possible to do so by replacing 
https://github.com/optuna/optuna/blob/01b7d148b1c0a7b2e5967ed40c6d0a337e05a7b3/optuna/integration/pytorch_lightning.py#L44

with `pl_module.global_step`. So if you would like to apply optuna to your use case as soon as possible, please write subclass `PyTorchLightningPruningCallback` and modify the definition of `on_validation_end` as I said. 

### As a feature request
Thank you for your suggestion!  Sometimes the mini-batch size is treated as hyper-parameters, which means that `global_step` can change depending on it. So unexpected comparison between global steps might happen; for example, when a learning scheduler changes `learning_rate` by epoch. I suppose we need to discuss the behaviour.

**Comment by ManuelAngel99** [comment]
Hi! 
Are there any workarounds this issue? I also need to call the validation step multiple times per epoch and was surprided when check_val_every_n_epoch didn't allow for non integer values
